bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java

```diff
@@ -461,87 +489,343 @@ public Boolean isEnabled(String ruleUID) {
+            }
         }
     }
 
     @Override
     public void added(Provider<Rule> provider, Rule element) {
-        super.added(provider, element);
-        updateRuleByTemplate(provider, element);
-    }
+        Rule ruleCopy = RuleUtils.getRuleCopy(element);
+        super.added(provider, ruleCopy);
+        postRuleAddedEvent(ruleCopy);
+        try {
+            Rule resolvedRule = resolveRuleByTemplate(ruleCopy);
+            if (resolvedRule.getTemplateUID() == null) {
+                RuleStatusInfo statusInfo = disabledRulesStorage != null
+                        && disabledRulesStorage.get(ruleCopy.getUID()) == null
+                                ? new RuleStatusInfo(RuleStatus.INITIALIZING)
+                                : new RuleStatusInfo(RuleStatus.DISABLED);
+                addToRuleEngine(resolvedRule, statusInfo);
 
-    private void updateRuleByTemplate(Provider<Rule> provider, Rule rule) {
-        Rule resolvedRule = resolveRuleByTemplate(rule);
-        if (rule != resolvedRule) {
-            if (update(resolvedRule) == null) {
-                super.updated(provider, rule, resolvedRule);
             }
+            if (ruleCopy != resolvedRule) {
+                update(resolvedRule);
+            }
+        } catch (Exception e) {
+            logger.debug("Added rule '{}' is invalid", ruleCopy.getUID(), e);
         }
     }
 
     @Override
     public void updated(Provider<Rule> provider, Rule oldElement, Rule element) {
-        Rule resolvedRule = resolveRuleByTemplate(element);
-        super.updated(provider, oldElement, resolvedRule);
+        String uid = element.getUID();
+        if (oldElement != null && uid.equals(oldElement.getUID())) {
+            try {
+                Rule resolvedRule = resolveRuleByTemplate(element);
+                super.updated(provider, oldElement, resolvedRule);
+                postRuleUpdatedEvent(resolvedRule, oldElement);
+                if (resolvedRule.getTemplateUID() == null) {
+                    RuleStatusInfo statusInfo = disabledRulesStorage != null && disabledRulesStorage.get(uid) == null
+                            ? new RuleStatusInfo(RuleStatus.INITIALIZING)
+                            : new RuleStatusInfo(RuleStatus.DISABLED);
+                    addToRuleEngine(resolvedRule, statusInfo);
+                }
+            } catch (Exception e) {
+                logger.error("The rule '{}' is not updated, the new version is invalid", uid, e);
+            }
+        } else {
+            throw new IllegalArgumentException(
+                    String.format("The rule '%s' is not updated, not matching with any existing rule", uid));
+        }
+    }
+
+    @Override
+    protected void onAddElement(Rule element) throws IllegalArgumentException {
+        try {
+            validateModuleIDs(element, element.getModules(Module.class));
+            resolveConfigurations(element);
+        } catch (Exception e) {
```

Comment:
We should not catch generic Exceptions. Can you make this more specific?

Reply:
OK